examples/NAT.cs

```diff
@@ -36,223 +36,348 @@
-    if (port_mapping.TryGetValue(from, out to))
+    InternalNode destination;
+    var key = new MapToInside_Key(p_ip.SourceAddress, p_tcp.SourcePort, p_tcp.DestinationPort);
+    if (NAT_MapToInside.TryGetValue(key, out destination))
     {
       // Rewrite destination IP address and TCP port, and map to the appropriate Inside port.
-      p_ip.DestinationAddress = to.ip_address;
-      p_tcp.DestinationPort = to.tcp_port;
-      // Update checksums.
-      p_tcp.UpdateTCPChecksum();
+      p_ip.DestinationAddress = destination.Address;
+      p_tcp.DestinationPort = destination.Port;
+      // Update checksums. NOTE IP updated before TCP
       ((IPv4Packet)p_ip).UpdateIPChecksum();
-      return to.network_port.Value;
+      p_tcp.UpdateTCPChecksum();
+      // Update destination MAC address
+      p_eth.DestinationHwAddress = destination.MacAddress;
+      // Forward on the mapped network port
+      return destination.NetworkPort;
     }
     else
     {
       return Port_Drop;
     }
   }
 
-  private int inside_to_outside (IpPacket p_ip, TcpPacket p_tcp, NAT_Entry from)
+  /// <summary>
+  /// Rewrite packets coming from the Iside and forward on the Outside network port.
+  /// </summary>
+  private int inside_to_outside (EthernetPacket p_eth, IpPacket p_ip, TcpPacket p_tcp, int incomingPort)
   {
-    // In this case, we're querying the dictionary with "reversed" keying,
-    // since we want to find out how the destination IP address and destination
-    // TCP port need to be rewritten before forwarding to the outside.
-
-    NAT_Entry to;
+    var out_key = new MapToOutside_Key(p_ip.SourceAddress, p_tcp.SourcePort,
+                                       p_ip.DestinationAddress, p_tcp.DestinationPort);
+    ushort natPort;
     if (p_tcp.Syn)
     {
-      // If a TCP SYN, then add a mapping.
-
-      // We first delete any previous mapping that had the same key.
-      if (port_reverse_mapping.TryGetValue(from, out to))
-      {
-        // We don't really care about the boolean result returned by TryRemove,
-        // since if "true" then the key-value pair was removed (great), but
-        // if "false" then the key-value pair must have been already removed (great).
-        port_mapping.TryRemove(port_reverse_mapping[from], out to);
-        port_reverse_mapping.TryRemove(from, out to);
-      } // FIXME in case TryGetValue returns false, we could simply do TryUpdate or TryAdd instead of the TryRemove and TryAdd.
-
-      to = new NAT_Entry();
-      to.ip_address = p_ip.DestinationAddress;
-      to.tcp_port = p_tcp.DestinationPort;
-
-      // Generate data for the mapping
-      lock(my_address)
-      {
-        // NOTE can't use Interlocked.Increment since the type of next_port is short not int.
-        to.assigned_tcp_port = next_port++;
-      }
+      // If a TCP SYN, then add a mapping for the new connection.
+      natPort = GetNATPort();
+      // Add to NAT_MapToOutside
+      NAT_MapToOutside[out_key] = natPort;
```

Comment:
Since we want to add or update the value regardless of any existing entry, we can just use the indexer. Under the hood, using the indexer has the same effect as using the TryAdd method (The [indexer](http://referencesource.microsoft.com/#mscorlib/system/Collections/Concurrent/ConcurrentDictionary.cs,52d16ab3fe5a7e2b) setter uses [TryAddInternal](http://referencesource.microsoft.com/#mscorlib/system/Collections/Concurrent/ConcurrentDictionary.cs,bc495056297a0851),  the same as [TryAdd](http://referencesource.microsoft.com/#mscorlib/system/Collections/Concurrent/ConcurrentDictionary.cs,2de2f6d77cb8d316)).